README.md
# UV-AADA-BigMart

Predict the sales

# How to test our project ? Either run tests/main_extract.py or tests/main_extract_bis.py

## tests/main_extract.py

We will run ./tests/main_extract.py which will use almost all our modules :
- load_data (loads a dataset from a csv file inside a dataframe)
- df_uniform_categorical (Uniforming the tags of a categorical feature colum, as Item_Fat_Content has many tag for one same thing)
- split_categorical_numerical (Split the categorical and numerical columns names)
- df_preprocessing_num_features_train_test(preprocessing the num features for df_train, df_test so that the null value take the mean value of the column of the df_train)
- transform_categorical_numerical_train_test (transform the categorical colums to numerical for the df_test and df_train.
  using LabelEncoder())
- regression__selection(Exhaustive search over specified parameters for a large pannel of regressor)

The main_explore.py in the tests folder was used to explore the data.
In the end it will save the best model and the transformed dataset(train and test).
The train dataset is used to train and test the model.

## tests/main_extract_bis.py

We will run ./tests/main_extract.py which will use almost all our modules, it's our test to select automaticaly our features :
- load_data
- df_uniform_categorical
- split_categorical_numerical
- df_preprocessing_num_features_train_test
- transform_categorical_numerical_train_test
- regression_and_features_selection(Exhaustive search over specified parameters for a large pannel of regressor and selecting features)

### install dependencies

#### if you are using python2

``` bash
$ pip install -r requirements.txt
```

#### if you are using python3

``` bash
$ pip3 install -r requirements.txt
```

### run our main test function

#### if you are using python2

``` bash
$ python main_name.py
```

#### if you are using python3

``` bash
$ python3 main_name.py
```

#### GridSearchCV result for main_extract:

```
Best params for each regressor

best_params_ for KNeighborsRegressor regressor
{'leaf_size': 15, 'n_neighbors': 9, 'weights': 'distance'}
best_score_ for KNeighborsRegressor regressor: 0.48036216829262707
R2 for KNeighborsRegressor best_score_ regressor: 0.5222367940146722
RMSE for KNeighborsRegressor best_score_ regressor: 1178.882890821993

best_params_ for RandomForestRegressor regressor
{'max_depth': None, 'max_features': 'auto', 'min_samples_leaf': 100, 'n_estimators': 400, 'n_jobs': 4, 'random_state': 0}
best_score_ for RandomForestRegressor regressor
0.5840023775597037
R2 for RandomForestRegressor best_score_ regressor: 0.5996242744010539
RMSE for RandomForestRegressor best_score_ regressor: 1079.1906748781348

best_params_ for CatBoostRegressor regressor
{'depth': 3, 'iterations': 100, 'learning_rate': 0.1, 'loss_function': 'RMSE'}
best_score_ for CatBoostRegressor regressor
0.5976648644070012
R2 for CatBoostRegressor best_score_ regressor: 0.6121388256861613
RMSE for CatBoostRegressor best_score_ regressor: 1062.1906364309425

best_params_ for GradientBoostingRegressor regressor
{'learning_rate': 0.03, 'max_depth': 4, 'n_estimators': 100, 'subsample': 0.5}
best_score_ for GradientBoostingRegressor regressor
0.5936869547812392
R2 for GradientBoostingRegressor best_score_ regressor: 0.6039579299923228
RMSE for GradientBoostingRegressor best_score_ regressor: 1073.3342194804886

best_params_ for DecisionTreeRegressor regressor
{'max_depth': 8, 'min_samples_leaf': 100}
best_score_ for DecisionTreeRegressor regressor
0.5825088678155786
R2 for DecisionTreeRegressor best_score_ regressor: 0.5961359345526586
RMSE for DecisionTreeRegressor best_score_ regressor: 1083.8817927869181

best_params_ for XGBRegressor regressor
{'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 500, 'n_jobs': 4}
best_score_ for XGBRegressor regressor
0.5958467441412338
R2 for XGBRegressor best_score_ regressor: 0.6090278050443259
RMSE for XGBRegressor best_score_ regressor: 1066.442024998566

best_params_ for LinearRegression regressor
{'normalize': True}
best_score_ for LinearRegression regressor
0.5069807272751782
R2 for LinearRegression best_score_ regressor: 0.512151655514078
RMSE for LinearRegression best_score_ regressor: 1191.2604747445223

best_params_ for Ridge regressor
{'alpha': 0.05, 'normalize': True}
best_score_ for Ridge regressor
0.5054256927204294
R2 for Ridge best_score_ regressor: 0.5104172373734344
RMSE for Ridge best_score_ regressor: 1193.3762046108932
```
#### Conclusion:

In order to find the best results, we used GridSearchCV to test a large number of regression models: (KNeighborsRegressor, RandomForestRegressor, CatBoostRegressor, GradientBoostingRegressor, DecisionTreeRegressor, XGBRegressor, LinearRegression, Ridge).

We found that the best regressor is CatBoostRegressor. Since its best iteration had the highest r2 and the lowest RMSE.

```
Results for CatBoostRegressor:
Best parameters used : {'depth': 3, 'iterations': 100, 'learning_rate': 0.1, 'loss_function': 'RMSE'}
Best R2 : 0.6121388256861613
Best RMSE : 1062.1906364309425
```

With main_extract_bis we didn't take enough time to tweak it, and only obtained the scores

```
RandomForestRegressor: 0.59966836436923
GradientBoostingRegressor: 0.6060770903781255
XGBRegressor: 0.6077281080442473
```

We haven't analysed the features that much, and we have't proceeded to a feature elimination. Eliminating certain features enables the machine learning algorithm to train faster, while reducing the complexity of a model and makes it easier to interpret. It also improves the accuracy of a model if the right subset is chosen. We could also avoid overfitting

With our main_explore.py function we can see in one of our graph that there is a relationship between the price(Item_MRP) and sales(Item_Outlet_Sales)

![img](docs/img/aichatoutoure-img1.png")

The scatterplot below also suggests that the price 'Item_MRP' and the 'Item_Outlet_Sales' are strongly correlated. We can conclude that Item_Outlet_Sales would be a good feature too keep

![img](docs/img/aichatoutoure-img2.png)

The bar below shows a big difference between the "Item_Outlet_Sales" between the different "Outlet_type".n We can also conclude that "Outlet_type" would be a good feature too keep.

![img](docs/img/aichatoutoure-img3.png)

# Predicting the Item_Outlet_Sales from treated test dataset:

## Using tests/main_.py

### install dependencies

#### if you are using python2

``` bash
$ pip install -r requirements.txt
```

#### if you are using python3

``` bash
$ pip3 install -r requirements.txt
```

### run our main test function

#### if you are using python2

``` bash
$ python main_.py
```

#### if you are using python3

``` bash
$ python3 main_.py
```

### Results:

```
[1734.87853431 1352.48154455  696.46521969 ... 1818.59764884 3551.4866275
 1161.2948966 ]
```

## workflows


## workflows

### semantic-version

[![semantic-version](https://img.shields.io/github/workflow/status/LazyKeru/UV-AADA-projet-apprentissage-automatique/Semantic-version?style=plastic)](https://github.com/LazyKeru/UV-AADA-BigMart/actions/workflows/semantic-versioning.yml)

### Run Python Tests
[![Run Python Tests](https://img.shields.io/github/workflow/status/LazyKeru/UV-AADA-projet-apprentissage-automatique/Semantic-version?style=plastic)](https://github.com/LazyKeru/UV-AADA-BigMart/actions/workflows/python-tests.yml)
#### test are runned for:
- ./project/function/load_data module with ./tests/test_load_data
- ./project/function/feature_extraction with ./tests/test_feature_extraction.py

## Commit convention :
- build: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
- ci: Changes to our CI configuration files and scripts (examples: CircleCi, SauceLabs)
- docs: Documentation only changes
- feat: A new feature
- fix: A bug fix
- perf: A code change that improves performance
- refactor: A code change that neither fixes a bug nor adds a feature
- test: Adding missing tests or correcting existing tests


docs/CHANGELOG.md
# [1.5.0](https://github.com/LazyKeru/UV-AADA-BigMart/compare/v1.4.0...v1.5.0) (2022-01-12)


### Features

* project should be ready to hand in ([208b9c6](https://github.com/LazyKeru/UV-AADA-BigMart/commit/208b9c627020d874bf7a43b91f4f102523cc684e))

# [1.4.0](https://github.com/LazyKeru/UV-AADA-BigMart/compare/v1.3.0...v1.4.0) (2022-01-12)


### Features

* added regression and feature selection module ([a1abe2e](https://github.com/LazyKeru/UV-AADA-BigMart/commit/a1abe2eae459e69d6e2a8a236ba03a38619f66af))


### Performance Improvements

* changed the description for each modules and cleaning the test section ([2e11660](https://github.com/LazyKeru/UV-AADA-BigMart/commit/2e11660ad661466a5a6688048138b44827d26d36))
* optimized the transform_categorical_numerical_train_test module ([34774ef](https://github.com/LazyKeru/UV-AADA-BigMart/commit/34774ef76d2568edd99fcf1230b6cc1bb067882c))

# [1.3.0](https://github.com/LazyKeru/UV-AADA-BigMart/compare/v1.2.1...v1.3.0) (2022-01-11)


### Features

* split analyse and regression selection ([74cea7e](https://github.com/LazyKeru/UV-AADA-BigMart/commit/74cea7e03e859ae433ef22f3320f5936e1720577))

## [1.2.1](https://github.com/LazyKeru/UV-AADA-BigMart/compare/v1.2.0...v1.2.1) (2022-01-11)


### Performance Improvements

* added new reg ([c8070fd](https://github.com/LazyKeru/UV-AADA-BigMart/commit/c8070fd72add4780a07a5a7d95a8e17e9d1cead2))

# [1.2.0](https://github.com/LazyKeru/UV-AADA-BigMart/compare/v1.1.0...v1.2.0) (2022-01-11)


### Features

* added df_preprocessing_num_features module ([345ac2d](https://github.com/LazyKeru/UV-AADA-BigMart/commit/345ac2d8b5fa35869da4d5ed9b86b13bc5768740))
* added randomForestRegressor ([eed643c](https://github.com/LazyKeru/UV-AADA-BigMart/commit/eed643ca5f8f12822aed786c330857bc886cf8d8))
* added the explore_categorical_features.py ([7d5a297](https://github.com/LazyKeru/UV-AADA-BigMart/commit/7d5a2979222c142deafd0c3827ec35e8a3536ede))
* added the explore_numerical_features.py ([8558b41](https://github.com/LazyKeru/UV-AADA-BigMart/commit/8558b41e60941303488c29b150890354aac129cc))
* df_uniform_categorical module added ([114f3a2](https://github.com/LazyKeru/UV-AADA-BigMart/commit/114f3a222dbcee1ce20ce8a2ac466fd7e85976cd))


### Performance Improvements

* transform_categorical_numerical updated ([291bb77](https://github.com/LazyKeru/UV-AADA-BigMart/commit/291bb776527e5fd9198de5ec0f0f9c07ac464ecf))

# [1.1.0](https://github.com/LazyKeru/UV-AADA-BigMart/compare/v1.0.0...v1.1.0) (2022-01-11)


### Features

* added feature_search module ([f9366b8](https://github.com/LazyKeru/UV-AADA-BigMart/commit/f9366b89cee3db279ee279e0020b712fc6a8c821))
* added transform_categorical_numerical module ([da735e6](https://github.com/LazyKeru/UV-AADA-BigMart/commit/da735e66b90017299f7de87081ecd77afb29df67))


### Performance Improvements

* made the load_data module more global ([a784a4c](https://github.com/LazyKeru/UV-AADA-BigMart/commit/a784a4c73c477042de165d85449a8e358b4ae6c9))

# 1.0.0 (2022-01-11)


### Features

* the data will now load ([8d401dd](https://github.com/LazyKeru/UV-AADA-BigMart/commit/8d401ddb1ce8f48441949af77ee5ce3da7b8d51b))


project/regression_selection.py
# classifier_parameters_selection
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
# to add more parameters to the default classifier
import numpy as np
# classifiers, remove if your remove default
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import RandomForestRegressor
from catboost import CatBoostRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.tree import DecisionTreeRegressor
from xgboost import XGBRegressor
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from .analyse import analyse

default_names = [
    #"LogisticRegression", # Neural network models
    "KNeighborsRegressor",
    "RandomForestRegressor",
    "CatBoostRegressor",
    "GradientBoostingRegressor",
    "DecisionTreeRegressor",
    "XGBRegressor",
    "LinearRegression",
    "Ridge"
]

default_classifiers = [
    #LogisticRegression(),
    KNeighborsRegressor(),
    RandomForestRegressor(),
    CatBoostRegressor(),
    GradientBoostingRegressor(),
    DecisionTreeRegressor(),
    XGBRegressor(),
    LinearRegression(),
    Ridge()
]

default_parameters = [
    #{
    #    'penalty':['l1', 'l2', 'elasticnet'],        # l1 is Lasso, l2 is Ridge
    #    'solver':['liblinear'],
    #    'C': np.linspace(0.00002,1,100)
    #},
    {
        'weights': ['uniform', 'distance'],
        'n_neighbors': [3, 5, 7, 9],
        'leaf_size': [15, 20],
        # Many more config : https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html
    },
    {
        'n_estimators': [100, 200, 400, 1000],
        'max_features': ["auto", "log2"],
        'max_depth': [None, 5, 6],
        'min_samples_leaf': [100],
        'n_jobs': [4, 5],
        'random_state': [0]
        # Many more config : https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html
    },
    {
        'iterations': [50, 100, 150, 200, 300],
        'depth': [3],
        'learning_rate': [0.1],
        'loss_function': ['RMSE']
        # Many more config : https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html
    },
    {
        'learning_rate': [0.01,0.02,0.03,0.04],
        'subsample'    : [0.9, 0.5, 0.2, 0.1],
        'n_estimators' : [100,500,1000, 1500],
        'max_depth'    : [4,6,8,10]
    },
    {
        'max_depth': [8, 16, 30],
        'min_samples_leaf'    : [100, 150, 200]
    },
    {
        'n_estimators': [50,100,200,500,1000],  # this parameter means using the GPU when training our model to speedup the training process
        'max_depth': [3,5,7,9],
        'n_jobs': [4],
        'learning_rate': [0.001,0.01,0.02,0.05,0.1],
    },
    {
        'normalize': [True]
    },
    {
        'alpha': [0.05, 0.1],
        'normalize': [True]
    }

]

def regression_selection(df, target, names=default_names,classifiers=default_classifiers,parameters=default_parameters):
    """
    Exhaustive search over specified parameters for a large pannel of regressor
    :param dataframe df: the data that we will split using train_test_split() for training
    :param array names: the names of the regressors
    :param array classifiers: the function of the regressors
    :param array parameters: the parameters of the regressors
    :return: it returns the best obtained regressor model
    """
    #splitting data
    features=[col for col in df.columns if col!=target]
    x_train, x_test, y_train, y_test = train_test_split(
        df[features],
        df[target],
        test_size=0.3,
        random_state=1)
    regs = {}
    for name, classifier, parameter in zip(names, classifiers, parameters):
        print(f"Start GridSearchCV for {name} classifier")
        #Exhaustive search over specified parameter for a specific classifier
        reg = GridSearchCV(classifier, parameter, scoring='r2', n_jobs=-1, cv=3)
        reg.fit(x_train, y_train)
        reg.predict(x_test)
        print(f"End GridSearchCV for {name} classifier")
        regs[name] = reg
        pass
    # sent to analyse function
    """
    for name, reg in regs.items():
        print(f"best_params_ for {name} classifier\n{reg.best_params_}")
        print(f"best_score_ for {name} classifier\n{reg.best_score_}")
        print(f'R2 for {name}: {r2_score(y_test, reg.best_estimator_.predict(x_test))}')
        print(f'RMSE for {name}: {np.sqrt((mean_squared_error(y_test, reg.best_estimator_.predict(x_test))))}')
        pass
    """
    best_r2_score=0
    for name, reg in regs.items():
        if best_r2_score < r2_score(y_test, reg.best_estimator_.predict(x_test)):
            best_r2_score = r2_score(y_test, reg.best_estimator_.predict(x_test))
            best_reg = reg.best_estimator_
            pass
        pass
    pass
    analyse(regs, x_test, y_test)
    return best_reg
